### PR TITLE
Add support for `Miniscope-DAQ-V3` and `MiniscopeAnalysis`.  Updated Jupyter Notebooks with detailed explanations.

### PR DESCRIPTION
**Pull request by kabilar**

### CODE REVIEW

**Review by dimitri-yatsenko — APPROVED** [review]

File: workflow_miniscope/__init__.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 __author__ = "DataJoint NEURO"
-__date__ = "December 15, 2020"
+__date__ = "May 1, 2021"
 __version__ = "0.0.1"
 
 __all__ = ['__author__', '__version__', '__date__']
```
Comment (ttngu207) [inline workflow_miniscope/__init__.py:5]: Let's put this information into a [`version.py`](https://github.com/ttngu207/workflow-array-ephys/blob/main/workflow_array_ephys/version.py) file, and modify the [`setup.py`](https://github.com/ttngu207/workflow-array-ephys/blob/main/setup.py) accordingly as well

File: workflow_miniscope/ingest.py [hunk 2]
```diff
@@ -24,43 +24,30 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
...
     for sess in input_sessions:
         sess_dir = pathlib.Path(sess['session_dir'])
 
-        # search for either ScanImage or ScanBox files (in that order)
-        for scan_pattern, scan_type, glob_func in zip(['*.tif', '*.sbx'],
-                                                      ['ScanImage', 'ScanBox'],
+        # Search for Miniscope-DAQ-V3 files (in that order)
+        for scan_pattern, scan_type, glob_func in zip(['*.avi'],
+                                                      ['Miniscope-DAQ-V3'],
                                                       [sess_dir.glob, sess_dir.rglob]):
```
Comment (ttngu207) [inline workflow_miniscope/ingest.py:36]: you probably want:
```
for scan_pattern, scan_type, glob_func in zip(['*.avi'],
                                                      ['Miniscope-DAQ-V3'],
                                                      [sess_dir.glob]):
```
(if this needs a recursive search, then use `sess_dir.rglob` instead)

File: workflow_miniscope/paths.py [hunk 3]
```diff
@@ -21,7 +21,7 @@ def get_scan_image_files(scan_key):
     if tiff_filepaths:
         return tiff_filepaths
     else:
-        raise FileNotFoundError(f'No tiff file found in {sess_dir}')
+        raise FileNotFoundError(f'No .tif file found in {sess_dir}')
 
 
 def get_scan_box_files(scan_key):
```
Comment (ttngu207) [inline workflow_miniscope/paths.py:27]: Should this be removed from the `miniscope` element/workflow entirely? 
(We can keep ScanImage, as that may still be used in many miniscope setup, but I don't think it's the case for `ScanBox`)

File: workflow_miniscope/ingest.py [hunk 4]
```diff
@@ -24,43 +24,30 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
...
     session_list, session_dir_list, scan_list, scanner_list = [], [], [], []
 
     for sess in input_sessions:
         sess_dir = pathlib.Path(sess['session_dir'])
 
-        # search for either ScanImage or ScanBox files (in that order)
-        for scan_pattern, scan_type, glob_func in zip(['*.tif', '*.sbx'],
-                                                      ['ScanImage', 'ScanBox'],
+        # Search for Miniscope-DAQ-V3 files (in that order)
+        for scan_pattern, scan_type, glob_func in zip(['*.avi'],
```
Comment (ttngu207) [inline workflow_miniscope/ingest.py:34]: perhaps `ms*.avi` to be more specific? (safeguard against other .avi files - e.g. tracking video)

**Review by ttngu207 — CHANGES_REQUESTED** [review]

File: workflow_miniscope/ingest.py [hunk 5]
```diff
@@ -24,43 +24,30 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
...
-                recording_time = datetime.fromtimestamp(sbx_fp.stat().st_ctime)  # read from file when scanbox support this
-                scanner = sbx_meta.get('imaging_system', 'ScanBox')
+            raise FileNotFoundError(f'Unable to identify scan files from the supported acquisition softwares (Miniscope-DAQ-V3) at: {sess_dir}')
+
+        if acq_software == 'Miniscope-DAQ-V3':
+            try:  # attempt to read Miniscope DAQ V3 file (.avi)
+                daq_v3_fp = pathlib.Path(scan_filepaths[0])
+                recording_time = datetime.fromtimestamp(daq_v3_fp.stat().st_ctime)
+                scanner = 'Miniscope-DAQ-V3'
             except Exception as e:
```
Comment (dimitri-yatsenko) [inline workflow_miniscope/ingest.py:49]: Does suppressing the error sufficiently alert the user of the problem? I think generally, errors should be raised to allow the calling context to handle the condition.

File: workflow_miniscope/paths.py [hunk 6]
```diff
@@ -39,3 +39,20 @@ def get_scan_box_files(scan_key):
...
+
+    from .pipeline import session
+    sess_dir = data_dir / (session.SessionDirectory & scan_key).fetch1('session_dir')
+
+    if not sess_dir.exists():
+        raise FileNotFoundError(f'Session directory not found ({sess_dir})')
+
+    miniscope_filepaths = [fp.as_posix() for fp in sess_dir.glob('ms*.avi')]
+    miniscope_filepaths.append((sess_dir / 'timestamp.dat').as_posix())
+    if miniscope_filepaths:
```
Comment (dimitri-yatsenko) [inline workflow_miniscope/paths.py:55]: this condition will always be true since the line above adds one element to it.

File: README.md [hunk 7]
```diff
@@ -13,6 +13,10 @@ This repository provides demonstrations for:
     + predefined directory lookup methods (see [workflow_miniscope/paths.py](workflow_miniscope/paths.py))
 3. Ingestion of clustering results (built-in routine from the imaging element)
 
++ This workflow provides support for miniscope calcium imaging data acquired with the `Miniscope-DAQ` systems.  And for data processed with `Suite2p`, `CaImAn`, or `Miniscope Analysis`.
+    + Note the [`Miniscope Analysis`](https://github.com/etterguillaume/MiniscopeAnalysis) package is not currently maintained.
```
Comment (shenshan) [inline README.md:17]: A naive question here: why do we support the Miniscope-DAQ-V3 and MiniscopeAnalysis in the workflow not in the element?

Comment (kabilar) [reply README.md:17]: Good question.  `Miniscope-DAQ-V3` and `MiniscopeAnalysis` are supported in both the `workflow` and `element`.  Metadata from `Miniscope-DAQ-V3` is ingested in the `scan` schema, and processed data from `MiniscopeAnalysis` is ingested in the `imaging` schema.

File: workflow_miniscope/paths.py (hunk 3 above)
Comment (kabilar) [reply workflow_miniscope/paths.py]: Good catch.  Removed all `Scanbox` support from the `element` and `workflow`.

File: workflow_miniscope/ingest.py (hunk 4 above)
Comment (kabilar) [reply workflow_miniscope/ingest.py]: Good idea.  Fixed.

File: workflow_miniscope/ingest.py (hunk 2 above)
Comment (kabilar) [reply workflow_miniscope/ingest.py]: Thanks for the explanation.  Fixed.

File: workflow_miniscope/__init__.py (hunk 1 above)
Comment (kabilar) [reply workflow_miniscope/__init__.py]: Good idea.  Updated both the `workflow` and `element` according to your example.

File: workflow_miniscope/paths.py (hunk 6 above)
Comment (kabilar) [reply workflow_miniscope/paths.py]: Good catch.  Fixed.

File: workflow_miniscope/ingest.py (hunk 5 above)
Comment (kabilar) [reply workflow_miniscope/ingest.py:49]: Upon further review, in this case we don't need the try/except.  If the files are missing, an error would be raised in the above for/else statement.  I have removed the try/catch statement.  Thanks.

**Comment by kabilar** [comment]
@shenshan @ttngu207 @dimitri-yatsenko

Thank you for your suggestions.  I have updated the pull request and tested the `workflow-miniscope` to manually verify data ingestion.

Please note the Jupyter Notebooks are now synced with `jupytext` so you can more easily read and comment on the scripts (.py), and ignore the notebooks (.ipynb).

Please review and let me know if you have any additional suggestions.  Thank you.

File: workflow_miniscope/paths.py [hunk 8]
```diff
@@ -34,8 +34,11 @@ def get_scan_box_files(scan_key):
...
         raise FileNotFoundError(f'Session directory not found ({sess_dir})')
 
-    sbx_filepaths = [fp.as_posix() for fp in sess_dir.glob('*.sbx')]
-    if sbx_filepaths:
-        return sbx_filepaths
+    avi_filepaths = [fp.as_posix() for fp in sess_dir.glob('ms*.avi')]
+    dat_filepaths = [fp.as_posix() for fp in sess_dir.glob('timestamp.dat')]
+    miniscope_filepaths = avi_filepaths + dat_filepaths
+
+    if avi_filepaths and dat_filepaths:
```
Comment (ttngu207) [inline workflow_miniscope/paths.py:41]:
```suggestion
    if miniscope_filepaths :
```

File: workflow_miniscope/paths.py [hunk 9]
```diff
@@ -34,8 +34,11 @@ def get_scan_box_files(scan_key):
     if not sess_dir.exists():
         raise FileNotFoundError(f'Session directory not found ({sess_dir})')
 
-    sbx_filepaths = [fp.as_posix() for fp in sess_dir.glob('*.sbx')]
-    if sbx_filepaths:
-        return sbx_filepaths
+    avi_filepaths = [fp.as_posix() for fp in sess_dir.glob('ms*.avi')]
+    dat_filepaths = [fp.as_posix() for fp in sess_dir.glob('timestamp.dat')]
+    miniscope_filepaths = avi_filepaths + dat_filepaths
```
Comment (ttngu207) [inline workflow_miniscope/paths.py:39]:
```suggestion
    avi_filepaths = [fp.as_posix() for fp in sess_dir.glob('ms*.avi')]
    dat_filepath = (sess_dir  / 'timestamp.dat').as_posix()
    miniscope_filepaths = avi_filepaths.append(dat_filepath )
```

File: workflow_miniscope/paths.py (hunk 8 above)
Comment (ttngu207) [inline workflow_miniscope/paths.py:41]:
```suggestion
    if avi_filepaths and dat_filepath.exists():
```

File: workflow_miniscope/ingest.py [hunk 10]
```diff
@@ -24,44 +24,27 @@ def ingest_sessions(session_csv_path='./user_data/sessions.csv'):
     with open(session_csv_path, newline= '') as f:
```
[the remark this answers is not in the document]
Comment (kabilar) [reply workflow_miniscope/ingest.py]: Thanks.  Fixed.

File: requirements.txt [hunk 11]
```diff
@@ -1,6 +1,8 @@
-datajoint>=0.13.dev7
-element-calcium-imaging @ git+https://github.com/datajoint/element-calcium-imaging.git
+datajoint>=0.13
+datajoint-connection-hub
```
[the remark this answers is not in the document]
Comment (kabilar) [reply requirements.txt:2]: Required for `djHub`.  Removed from this repository.  Will add back as necessary for the validation sites.

File: README.md [hunk 12]
```diff
@@ -138,12 +144,12 @@ however, in this particular `workflow-miniscope`, we take the assumption that th
     
 + Each `session` directory should contain:
  
-    + All `.tif` files for the scan, with any naming convention
+    + All `.tif` or `.sbx` or `.json` files for the scan, with any naming convention
```
[the remark this answers is not in the document]
Comment (kabilar) [reply README.md]: Good catch.  Fixed.

File: workflow_miniscope/paths.py (hunk 9 above)
Comment (kabilar) [reply workflow_miniscope/paths.py]: Thanks.

File: workflow_miniscope/paths.py (hunk 8 above)
Comment (kabilar) [reply workflow_miniscope/paths.py]: Thanks.

**Comment by kabilar** [comment]
Thanks @ttngu207.  I have implemented your suggestions.

**Review by ttngu207 — APPROVED** [review]

File: notebooks/01ingest.py [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+                       scan_id=0, 
+                       scanner='Miniscope-DAQ-V3', 
+                       acq_software='Miniscope-DAQ-V3',
+                       scan_notes=''))
+
+# ## Populate `scan.ScanInfo`
+#
+# + This imported table stores information about the acquired image (e.g. image dimensions, file paths, etc.).
+
+populate_settings = {'display_progress': True, 'reserve_jobs': False, 'suppress_errors': False}
```
Comment (shenshan) [inline notebooks/01ingest.py:80]: should we remove `reserve_jobs` and `suppress_errors`? seems irrelevant to context here. And let's explain what `populate_settings` and `display_progress` do

File: notebooks/02explore.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+# # [WIP] DataJoint U24 Workflow Imaging
+# This notebook will describe the steps for interacting with the data ingested into `workflow-miniscope`.  
+#
+# Prior to using this notebook, please refer to the [README](https://github.com/datajoint/workflow-imaging) for the topics listed below.  
+#     + Installation instructions  
+#     + Directory structure and file naming convention  
+#     + Running the workflow  
+
+import os
+os.chdir('..')
```
Comment (shenshan) [inline notebooks/02explore.py:27]: explain why we do this (to find the json config)

File: notebooks/03automate.py [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+#       jupytext_version: 1.11.1
+#   kernelspec:
+#     display_name: 'Python 3.7.10 64-bit (''workflow-miniscope'': conda)'
+#     metadata:
+#       interpreter:
+#         hash: fb69438121b2de1073146525f02c3c959a56b9d2c1d6b0a62c34f01201323d1c
+#     name: python3
+# ---
+
+# # [WIP] Automated workflow
```
Comment (shenshan) [inline notebooks/03automate.py:18]: In the next round, let's add some descriptions on what this notebook does and the context when to use this notebook.

File: notebooks/04drop.py [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+#     display_name: 'Python 3.7.10 64-bit (''workflow-miniscope'': conda)'
+#     metadata:
+#       interpreter:
+#         hash: fb69438121b2de1073146525f02c3c959a56b9d2c1d6b0a62c34f01201323d1c
+#     name: python3
+# ---
+
+# # Drop schemas
+#
+# + This notebook is NOT required.
```
Comment (shenshan) [inline notebooks/04drop.py:20]: This is a little confusing. We could say something like "Schemas are not typically get dropped in a production pipeline with real data in it. At the developmental phase, it might be required for the table redesign. When dropping all schemas is needed, the following is the dependency order."

Comment (kabilar) [reply notebooks/04drop.py]: Great explanation.  Updated this notebook.

File: notebooks/02explore.py (hunk 14 above)
Comment (kabilar) [reply notebooks/02explore.py]: Thanks.  Added the explanation.

File: notebooks/01ingest.py (hunk 13 above)
Comment (kabilar) [reply notebooks/01ingest.py]: Thanks.  I have removed those two arguments and added explanations.

**Comment by kabilar** [comment]
Thanks @shenshan.  I have implemented your suggestions.
